signer/src/stacks/contracts.rs

```diff
@@ -77,14 +93,43 @@ pub trait AsContractCall {
     /// post-conditions, which is the address that sent the asset. The
     /// default is that we do not enforce any conditions since we usually
     /// deployed the contract.
-    fn post_conditions(&self, _: StacksAddress) -> StacksTxPostConditions {
+    fn post_conditions(&self) -> StacksTxPostConditions {
         StacksTxPostConditions {
             post_condition_mode: TransactionPostConditionMode::Allow,
             post_conditions: Vec::new(),
         }
     }
 }
 
+/// A generic newtype that implements AsTxPayload for all types that
+/// implement AsContractCall.
+///
+/// # Notes
+///
+/// Ideally, every type that implements AsContractCall should implement
+/// AsTxPayload automatically. What we want is to have something like the
+/// following:
+///
+/// impl<T: AsContractCall> AsTxPayload for T { ... }
+///
+/// But that would preclude us from adding something like:
+///
+/// impl<T: AsSmartContract> AsTxPayload for T { ... }
+///
+/// since doing so is prevented by the compiler because it introduces
+/// ambiguity. One work-around is to use a wrapper type that implements the
+/// trait that we want.
+pub struct ContractCall<T: AsContractCall>(pub T);
```

Comment:
I also guess we should proactively derive any relevant traits according to https://rust-lang.github.io/api-guidelines/interoperability.html
```suggestion
/// trait that we want.
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ContractCall<T>(pub T);
```

Since this is a newtype we probably also want `From` conversions between the wrapped and inner type, as well as `Deref` and `DerefMut`.

It might be overkill for now to implement everything, but I also think it's nice to have many of these traits in place since it can make it much more ergonomic to work with the wrapper.

Reply:
Couldn't hurt, and it's easy to do. Will add.